# Copyright 2026 The OpenXLA Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
Required to avoid crash due to missing encoding on convert layout ops.
This can happen when warp specialization is enabled, which calls into OptimizePartitionWarps,
which resets all layouts during its execution.
--- a/lib/Dialect/TritonGPU/IR/Ops.cpp
+++ b/lib/Dialect/TritonGPU/IR/Ops.cpp
@@ -67,6 +67,8 @@ bool isConvertTrivial(ConvertLayoutOp op) {
   auto dstType = op.getType();
   auto srcEncoding = srcType.getEncoding();
   auto dstEncoding = dstType.getEncoding();
+  if (!srcEncoding || !dstEncoding)
+    return false;
   return cast<DialectInferLayoutInterface>(&srcEncoding.getDialect())
       ->verifyLayoutsAreEqual(srcType.getShape(), srcEncoding, dstEncoding, {})
       .succeeded();

